### PR TITLE
feat: add async rest transport support in gapics

### LINKED ISSUE #2152: Add Support for different api methods for asynchronous rest transport.
Add support for the following types of api methods in asynchronous rest transport:

- LROs
- Extended LROs
- Pagers
- Server Streaming

Follow up:
- As we add support for each of these api methods, we need to updated the `is_rest_unsupported_method` macro in `test_macros.j2` file which guards the generation of tests for any of the above api methods.
- search for this issue number / url in the code and update the code wherever it's referenced.

Note: There are no plans of supporting client streaming api methods for sync or async rest transport.

### PR DESCRIPTION
**Pull request by ohmayr**
This PR implements experimental support for asynchronous rest transport in gapics.

Note: Changes should not be pushed directly to to `async-rest-support-in-gapics`. Once the work is completed, this PR can be merged.

Fixes: https://github.com/googleapis/gapic-generator-python/issues/2152

### CODE REVIEW

**Comment by ohmayr** [comment]
I had to manually update the goldens after resolving merge conflicts. This PR can be merged once we do a release in the following order:

- release `google-auth`
- release `api-core`
- cleanup noxfile.py to use the packages from pypi.
- merge this PR in main.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest_base.py.j2 [hunk 1]
```diff
@@ -66,7 +65,12 @@ class _Base{{ service.name }}RestTransport({{service.name}}Transport):
         Args:
             host ({% if service.host %}Optional[str]{% else %}str{% endif %}):
                 {{ ' ' }}The hostname to connect to {% if service.host %}(default: '{{ service.host }}'){% endif %}.
-            credentials (Optional[google.auth.credentials.Credentials]): The
+            {# Note: Type hint for credentials is set to `Any` to support
+             # async and sync credential types in the parent rest transport classes. 
+             # However, we should have a stronger type here such as an abstract base credentials
```
Comment (parthea) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest_base.py.j2:70]: Do we need a TODO comment/bug to follow up

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 2]
```diff
@@ -1053,10 +1066,13 @@ def test_transport_adc(transport_class):
 {% endfor %}
 {% for conf in configs %}
 {{ test_macros.transport_kind_test(**conf) }}
-
+{{ test_macros.run_transport_tests_for_config(**conf) }}
 {{ test_macros.empty_call_test(**conf) }}
 {% endfor %}
-
+{# TODO(https://github.com/googleapis/gapic-generator-python/issues/2121): Remove the below macro when async rest is GA. #}
```
Comment (parthea) [inline gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:1072]: Is this TODO comment for the if statement or the actual macro?

**Review by parthea — APPROVED** [review]

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest_base.py.j2 (hunk 1 above)
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest_base.py.j2:70]: addressed.

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 (hunk 2 above)
Comment (ohmayr) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2]: addressed by moving the TODO comment next to the macro.

**Pull request merged**